### PR TITLE
Systematize documentation of environment variables

### PR DESCRIPTION
**Pull request by greatroar**
What does this PR change? What problem does it solve?
-----------------------------------------------------

Cleans up the documentation of restic backup's environment variables.

Cache locations were documented inconsistently in three places.

The backup docs mentioned PATH being used to find fusermount, which is never run by restic backup, while not mentioning rclone and ssh. 

The notion of a "system-wide" environment variable makes no sense, so I 
put everything in one table.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Follow-up to #2769.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added documentation for the changes (in the manual)
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: doc/040_backup.rst [hunk 1]
```diff
@@ -453,12 +459,11 @@ environment variables. The following lists these environment variables:
...
-In addition to restic-specific environment variables, the following system-wide environment variables
-are taken into account for various operations:
+See :ref:`caching` for the rules concerning cache locations.
 
- * ``$XDG_CACHE_HOME/restic``, ``$HOME/.cache/restic``: :ref:`caching`.
- * ``$TMPDIR``: :ref:`temporary_files`.
- * ``$PATH/fusermount``: Binary for ``restic mount``.
+The external programs that restic may execute include ``rclone`` (for rclone
+backends) and ``ssh`` (for the SFTP backend). These may respond to further
+environment variables; see their respective manuals.
```
Comment (rawtaz) [inline doc/040_backup.rst:466]: Really good addition!

File: doc/040_backup.rst [hunk 2]
```diff
@@ -416,6 +416,12 @@ environment variables. The following lists these environment variables:
     RESTIC_CACHE_DIR                    Location of the cache directory
     RESTIC_PROGRESS_FPS                 Frames per second by which the progress bar is updated
 
+    HOME                                Backup location for caches
+    LocalAppData                        Backup location for caches
+    PATH                                Lookup path for external programs
+    TMPDIR                              Location for temporary files
+    XDG_CACHE_HOME                      Backup location for caches
```
Comment (rawtaz) [inline doc/040_backup.rst:423]: I think that this is out of scope for this documentation. These are well established standard variables/paths, that doesn't belong in a list of environment variables that is more specific to restic's workings. Also, most of them are already mentioned/described in the caching section. I suggest we don't include this change, what do you think @fd0 and @MichaelEischer?

**Review by rawtaz — CHANGES_REQUESTED** [review]
I added some comments :)

File: doc/cache.rst [hunk 3]
```diff
@@ -20,6 +19,9 @@ a lower version number is found the cache is recreated with the current
 version. If a higher version number is found the cache is ignored and left as
 is.
 
+The rest of this chapter describes the data structures for the local cache with
+version 1.
```
[the remark this answers is not in the document]
Comment (greatroar) [reply doc/cache.rst:23]: This was already here, I mostly moved it around (and rephrased it a tiny bit).

File: doc/manual_rest.rst [hunk 4]
```diff
@@ -398,9 +398,13 @@ This allows faster operations, since meta data does not need to be loaded from
 a remote repository. The cache is automatically created, usually in an
 OS-specific cache folder:
 
- * Linux/other: ``~/.cache/restic`` (or ``$XDG_CACHE_HOME/restic``)
- * macOS: ``~/Library/Caches/restic``
+ * macOS: ``$HOME/Library/Caches/restic``
  * Windows: ``%LOCALAPPDATA%/restic``
+ * Linux/other: ``$XDG_CACHE_HOME/restic``, or ``$HOME/.cache/restic`` if
+   ``XDG_CACHE_HOME`` is not set
```
[the remark this answers is not in the document]
Comment (greatroar) [reply doc/manual_rest.rst:404]: I found it rather odd to have "other" as the first bullet point. I get restore the ~ if you feel strongly about it :)

File: doc/040_backup.rst (hunk 2 above)
Comment (greatroar) [reply doc/040_backup.rst:423]: Knowing where data goes is crucial when operating a security-focused program. The fact that these are all customary doesn't imply that restic uses them, rather than using a custom RESTIC_TEMPDIR, or /dev/shm, or /run/$uid/restic, or anonymous files in its working directory.

As a practical example, if your homedir is encrypted but the rest of the disk is not, it's important to know that and to know exactly how to keep restic from writing outside the homedir. (I think restic only writes encrypted data to TMPDIR, but I'm not 100% sure.)

Finally, restic has actually used different combinations of variables for locating the cache in previous releases, so the variables aren't that well-established...

Comment (rawtaz) [reply doc/040_backup.rst:423]: I'm certainly not convinced. Please consider the following:

- The variables that are relevant for caches are mentioned in the caching section, and shouldn't be in here because that is redundant and these types of variables are not the goal of this list. If there's a need to describe that e.g. $HOME is a "backup location" for caches, then that should be described in the caching section which already talks about this type of information.

- $PATH is just too common to justify mentioning here. If we're going to mention such common variables, there are others to mention as well, and it's not a good idea. Same for $TMPDIR which is a common one too.

File: doc/040_backup.rst [hunk 5]
```diff
@@ -453,12 +459,11 @@ environment variables. The following lists these environment variables:
 
     RCLONE_BWLIMIT                      rclone bandwidth limit
 
-In addition to restic-specific environment variables, the following system-wide environment variables
-are taken into account for various operations:
+See :ref:`caching` for the rules concerning cache locations.
```
Comment (MichaelEischer) [inline doc/040_backup.rst:462]: This should mention, that this is about the default cache location. If `RESTIC_CACHE_DIR` is set, then that has precedence.

File: doc/040_backup.rst (hunk 2 above)
Comment (MichaelEischer) [reply doc/040_backup.rst:423]: Listing 3 (or actually 4) variables related to the cache location will probably confuse more that it helps. Regarding the `PATH` variable I think that it is that common that the documentation should only mention when it is _not_ used (e.g. if restic had hardcoded absolute paths).

`TMPDIR` on the other hand is useful to know as that way a user can e.g. direct it to a ramdisk.

File: doc/manual_rest.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (greatroar) [reply doc/manual_rest.rst:404]: Ok.

File: doc/040_backup.rst (hunk 2 above)
Comment (greatroar) [reply doc/040_backup.rst:423]: I removed everything except TMPDIR.

File: doc/040_backup.rst (hunk 5 above)
Comment (greatroar) [reply doc/040_backup.rst:462]: Changed to:

    See :ref:`caching` for the rules concerning cache locations when
    ``RESTIC_CACHE_DIR`` is not set.

**Comment by greatroar** [comment]
The test failure is unrelated, but maybe worth the trouble of tracking down:

```
--- FAIL: TestArchiverAbortEarlyOnError (0.02s)
    --- FAIL: TestArchiverAbortEarlyOnError/#01 (0.01s)
        testing.go:30: using low-security KDF parameters for test
        config.go:65: disabling check of the chunker polynomial
        archiver_test.go:1918: chdir to /tmp/restic-test-143559427
        archiver_test.go:1946: Snapshot return error: test error
        archiver_test.go:1948: track fs: map[.:1 dir:1 dir/file1:1 dir/file2:1 dir/file3:1 dir/file4:1 dir/file5:1 dir/file6:1 dir/file7:1]
        archiver_test.go:1952: opened dir/file7 1 times, want 0
        archiver_test.go:1955: chdir back to /home/travis/gopath/src/github.com/restic/restic/internal/archiver
FAIL
```

File: doc/cache.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (rawtaz) [reply doc/cache.rst:23]: @greatroar We discussed this a bit and would like to remove the entire Versions heading and its paragraphs, except the "Each repository has its own cache sub-directory, consisting of the repository ID
which is chosen at ``init``. All cache directories for different repos are
independent of each other." one which should be kept (instead of the Versions heading).

File: doc/040_backup.rst [hunk 6]
```diff
@@ -453,12 +459,11 @@ environment variables. The following lists these environment variables:
...
 
- * ``$XDG_CACHE_HOME/restic``, ``$HOME/.cache/restic``: :ref:`caching`.
- * ``$TMPDIR``: :ref:`temporary_files`.
- * ``$PATH/fusermount``: Binary for ``restic mount``.
```
[the remark this answers is not in the document]
Comment (greatroar) [reply doc/040_backup.rst:461]: Done by cherry-picking a commit from #2833 and adding a mention of fusermount.

File: doc/cache.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (greatroar) [reply doc/cache.rst]: Ok, done.

**Review by rawtaz — APPROVED** [review]
Bewm!
